aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -924,21 +965,67 @@ void TensorIterator::fast_set_up() {
       op.stride_bytes[0] = element_size_in_bytes;
     }
   }
+  return true;
 }
 
-bool TensorIterator::can_use_fast_set_up() {
-  if (is_reduction_) return false;
-  if (!all_ops_same_shape_) {
-    return false;
+FastSetupType TensorIterator::compute_fast_setup_type() {
+  if (is_reduction_ || !all_ops_same_shape_) {
+    return FastSetupType::NONE;
   }
-  for (auto& op : operands_) {
+
+  bool is_contiguous = true;
+  bool is_channels_last = true;
+  bool is_non_overlapping_and_dense = true;
+  for (const auto& op : operands_) {
     if (op.tensor.defined()) {
-      if (!op.tensor.is_contiguous()) {
-        return false;
+      is_contiguous &= op.tensor.is_contiguous(at::MemoryFormat::Contiguous);
+      is_channels_last &= op.tensor.is_contiguous(at::MemoryFormat::ChannelsLast);
+      is_non_overlapping_and_dense &= op.tensor.is_non_overlapping_and_dense();
+    }
+  }
+
+  if (is_contiguous) {
+    return FastSetupType::CONTIGUOUS;
+  }
+  if (is_channels_last) {
+    if (requires_channels_last_output_) {
+      return FastSetupType::CHANNELS_LAST;
+    }
+    return FastSetupType::NONE;
+  }
+  if (is_non_overlapping_and_dense) {
+    int64_t prev = -1;
+    bool restride_output = false;
+    // iterate from back to favor input's strides since we will restride the output tensor
+    // if only output's stride is not the same as input and all inputs' stride are the same
+    for (int64_t i = ntensors() - 1; i >= 0; --i) {
+      const auto& op = operands_[i];
+      if (op.tensor.defined()) {
+        if (prev < 0) {
+          prev = i;
+          continue;
+        }
+        if (!operands_[prev].tensor.strides().equals(op.tensor.strides())) {
+          if (!resize_outputs_ || !op.is_output || op.is_read_write) {
+            return FastSetupType::NONE;
+          }
+          if (!restride_output) {
+            restride_output = true;
+          }
+        }
+      }
+    }
+    if (restride_output) {
```

Comment:
compute_fast_setup_type feels like static function, until it is really updates output, this is bad for readability, perhaps you can restride outputs in fast_set_up

Reply:
Move restride into fast_set_up